### PR TITLE
New lint: (maybe_)infinite_iter

### LINKED ISSUE #1870: Lint (possible) infinite iteration
We can detect some infinite iterators (and some non-infinite):
* iter::repeat(_),Iterator::cycle(self), (_..) are always infinite
* Iterator::zip(self, _) is infinite if both arguments are
* Iterator::chain(self, _) is infinite if one of the arguments are
* Iterator::map(self, _), Iterator::by_ref(self), Iterator::cloned(self), Iterator::rev(self), Iterator::inspect(self) Iterator::enumerate(self), Iterator::peekable(self, _), Iterator::fuse(self), Iterator::skip(self, _), Iterator::skip_while(self, _), Iterator::filter(self, _), Iterator::filter_map(self, _), Iterator::flat_map(self, _), Iterator::unzip(self), Iterator::rev(self) are infinite if self is
* Iterator::take_while(self, _), Iterator::scan(self, _) are potentially infinite if self is
* Iterator::take(self, _) is always finite

The following methods do not complete iteration (and are thus safe to use even with infinite iterators):

* .next()
* .nth(_)

The following methods may potentially complete iteration, resulting in a possible infinite loop when called on infinite iterators:

* .find(_)
* .rfind(_)
* .position(_)
* .rposition(_)
* .any(_)
* .all(_)
* .comparisons / equality

The following methods *will* complete iteration, resulting in a guaranteed infinite loop in the absence of panics or aborts:

* .count()
* .collect()
* .fold(..)
* .for_each(_)
* .partition(_)
* .last() (unless `self : DoubleEndedIterator`)
* .max() / .max_by(_) / .max_by_key(_)
* .min() / .min_by(_) / .min_by_key(_)
* .sum()
* .product()

Also `for` loops over infinite iterators without `break`, `return` statement or `?` expression will loop infinitely.

I'd suggest having a two-tier lint:
* infinite_iteration (for guaranteed infinite iteration, warn by default)
* possible_infinite_iteration (for possibly unbounded infinite iteration because either the iterator or collector may abort iteration, allow by default)

### PR DESCRIPTION
**Pull request by llogiq**
This fixes #1870 (mostly, does not account for loops yet)

### CODE REVIEW

File: clippy_lints/src/infinite_iter.rs [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+}
+
+impl From<bool> for TriState {
+    fn from(b: bool) -> Self {
+        if b { True } else { False }
+    }
+}
+
+#[derive(Copy, Clone)]
+enum Heuristic {
```
Comment (oli-obk) [inline clippy_lints/src/infinite_iter.rs]: Can you document the variants a little?

File: clippy_lints/src/infinite_iter.rs [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+    Always,
+    First,
+    Any,
+    All
+}
+
+use self::Heuristic::{Always, First, Any, All};
+
+// here we use the `TriState` as (Finite, Possible Infinite, Infinite)
```
Comment (killercup) [inline clippy_lints/src/infinite_iter.rs]: As this is the only use case for TriState, I'd rename it to Finitude and the variants to Finite, PossiblyInfinite, and Infinite.

File: clippy_lints/src/infinite_iter.rs [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+static HEURISTICS : &[(&str, usize, Heuristic, TriState)] = &[
```
Comment (killercup) [inline clippy_lints/src/infinite_iter.rs]: That took me while to grasp. I'd turn that tuple into a `struct { method: &str, args: usize, heuristic: Heuristic, finitude: Finitude}`. Or add a doc comment. Same below for the other tuples, but it's not so important there.

File: clippy_lints/src/infinite_iter.rs (hunk 1 above)
Comment (llogiq) [reply clippy_lints/src/infinite_iter.rs]: Will do.

File: clippy_lints/src/infinite_iter.rs (hunk 2 above)
Comment (llogiq) [reply clippy_lints/src/infinite_iter.rs]: I had that initially, but renamed it to be more general when I added the second function, which would have used yet another tristate before I simplified the whole thing.

Perhaps we'll use a general TriState in the future, but for now I'll rename.

File: clippy_lints/src/infinite_iter.rs (hunk 3 above)
Comment (llogiq) [reply clippy_lints/src/infinite_iter.rs]: I'll rather go with the doc comment to avoid duplication.

**Comment by oli-obk** [comment]
Thanks. Those docs made the code much better to comprehend

**Comment by llogiq** [comment]
Yeah, thanks @oli-obk and @killercup for the review.

**Pull request merged**